File: cs50/ProblemSet2/caesar/caesar.c
#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>

const int UPPER_A_POSITION=65;
const int LOWER_A_POSITION=97;
const int CHECK_ARGUMENT=2;
const int NUMBER_LETTERS=26;

//==================================================================================================================================
int checkKeyInput(string input,int arg_number){
    int answer=0;
    int i=0;
    if (arg_number!=CHECK_ARGUMENT){
        answer=1;
    }
    else{
        while(input[i]!=0){
              if(!isdigit(input[i++])){
                 answer=1;
              }
        }
    }
    return answer;
}
//==================================================================================================================================
string getPlainText(){
       printf("plaintext:");
       return get_string();
}
//==================================================================================================================================
int checkModification(char letter){
    int mod=0;
    if(isupper(letter)){
        mod=UPPER_A_POSITION;
    }else{
        mod=LOWER_A_POSITION;
    }
    return mod;
}
//==================================================================================================================================
int applyCaesarCipher(char letter, int key){
    return((letter-checkModification(letter)+key)%NUMBER_LETTERS+checkModification(letter));
}
//==================================================================================================================================
void printCipherText(string plain_text, int cipher_key){
     int i=0;
     printf("ciphertext:");
     while(plain_text[i]!=0){
           if(isalpha(plain_text[i])){
               printf("%c",applyCaesarCipher(plain_text[i],cipher_key));
           }else{
               printf("%c",plain_text[i]);  
           }
           i++;
     }
     printf("\n");
}
//==================================================================================================================================
int main(int argc, string argv[]){
  if(checkKeyInput(argv[1],argc)){
     printf("Usage: ./caesar k\n"); 
     return 1;
  }
  printCipherText(getPlainText(),atoi(argv[1]));
  
  return 0;
}
//==================================================================================================================================

File: cs50/golang/caesar/caesar.go
package main

import (
	"fmt"
	"os"
	"unicode"
	"strconv"
	"bufio"
)

const NUMBER_OF_LETTERS =26
const LOWER_A_POSITION  = 97
var (
	input_text string
)
//======================================================================================================================
func applyCipherCaesar(letter rune, caesar_key int)rune{
	upper :=unicode.IsUpper(letter)
	letter=unicode.ToLower(letter)
	caesar_key%=NUMBER_OF_LETTERS
	letter=(letter-LOWER_A_POSITION+rune(caesar_key))%NUMBER_OF_LETTERS+LOWER_A_POSITION
	if upper{
		letter=unicode.ToUpper(letter)
	}
	return letter
}
//======================================================================================================================
func printCipherText(cipher_key int, plain_text string){
	fmt.Println("ciphertext")
	for _, one_char:=range plain_text{
		if unicode.IsLetter(one_char){
			fmt.Printf("%c",applyCipherCaesar(one_char,cipher_key))
		}else {
			fmt.Printf("%c",one_char)
		}
	}
	fmt.Print("\n")
}
//======================================================================================================================
func checkArguments()int {
	if  len(os.Args[1:])==0{
		fmt.Println("Usage: ./caesar key")
		os.Exit(1)
	}
	argv:=os.Args[1]
	for _,one_char:=range argv{
		if !unicode.IsDigit(one_char){
			fmt.Println("Key must be integer")
			os.Exit(2)
		}
	}
	cipher_key,err:=strconv.Atoi(argv)
	if err!=nil{
		fmt.Println("Error key converting")
		os.Exit(3)
	}
	return cipher_key
}
//======================================================================================================================
func main(){
	printCipherText(checkArguments(),
		func ()string{
			fmt.Println("plaintext:")
			reader := bufio.NewReader(os.Stdin)
			input_text, err := reader.ReadString('\n')
			if err!=nil{
				panic("you must enter text")
			}
			return input_text
		}())
}
//======================================================================================================================